### PR TITLE
Core/Packet: cleanup in bytebuffer

### PR DESCRIPTION
**Pull request by Viste**
## CHANGES PROPOSED:
-  cleanup ByteBuffer
-  validate every client packet through utf8 ( thnaks Shauren )
- bit ace delete :D
-  Add utf8limit helper function for truncating UTF8 strings by bytes count without corruption(Warlockbugs )
-  cleanup in utf8length
-  limit NormalizePlayerName by  MAX_INTERNAL_PLAYER_NAME
- other cleanups in utf8handle code

## ISSUES ADDRESSED:
- Closes 

## TESTS PERFORMED:
- Build
- basic ingame test(yell(on screenshot) say(on screenshot) whispers guild) shows thats nothing is break but need test hack stuff

## HOW TO TEST THE CHANGES:
- play WoW because its affects to packets
- send messages via all possible ways
## KNOWN ISSUES AND TODO LIST:
- [x] test in game
- [ ] test with hacktool
## Target branch(es):
- [x] Master

## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

File: src/server/game/Handlers/ChatHandler.cpp [hunk 1]
```diff
@@ -171,7 +171,8 @@ void WorldSession::HandleMessagechatOpcode(WorldPacket & recvData)
 	            if (sWorld->getBoolConfig(CONFIG_CHATLOG_ADDON))
 	            {
 		            std::string to, msg;
-		            recvData >> to >> msg;
+		            recvData >> to;
+                    msg = recvData.ReadCString();
```
Comment (Nefertumm) [inline src/server/game/Handlers/ChatHandler.cpp]: tabs

Comment (Viste) [reply src/server/game/Handlers/ChatHandler.cpp]: yes

**Comment by Viste** [comment]
Снимок экрана от 2020-03-19 00-54-33
Снимок экрана от 2020-03-19 00-54-53

**Comment by Viste** [comment]
basic ingame test shows thats nothing is break but need test hack stuff

**Comment by pklloveyou** [comment]
After testing, it works well，But don't know how to test a hack

**Comment by Helias** [comment]
@Viste did you merge any hack? if you know how tell us.
I will try to make some random tests sending malformed packets and playing normally WoW asap.
If we do not know other ways to test this I think we can merge after some general tests.

**Comment by Viste** [comment]
all hacks dont run on linux idk what do to with this

**Comment by pklloveyou** [comment]
I removed #2739, tested. My client crashed.

**Comment by pklloveyou** [comment]
I'll try again after removing the auction house PR #2684

**Comment by pklloveyou** [comment]
image

**Comment by Viste** [comment]
> I removed #2739, tested. My client crashed.
> I'll try again after removing the auction house PR #2684

> The server crashed,

i dont understand nothing O_O

**Comment by pklloveyou** [comment]
I removed those two fixes, testing this PR.I'm not sure if that's the right thing to do.、

**Comment by Viste** [comment]
> I removed those two fixes, testing this PR.I'm not sure if that's the right thing to do.、

you dont need to remove them

**Comment by pklloveyou** [comment]
But I only have these two WPS hack.

**Comment by pklloveyou** [comment]
Maybe, not the same kind of vulnerability to test?？？

**Comment by Viste** [comment]
that fix now only for chat packets but can be use for others

**Comment by pklloveyou** [comment]
To understand the，You'll have to test it with the new WPS hack，Sorry, I can't do that.

**Comment by Viste** [comment]
updated with changes from #2721

File: src/common/Utilities/Util.cpp [hunk 2]
```diff
@@ -541,19 +549,24 @@ void utf8printf(FILE* out, const char *str, ...)
...
-    size_t temp_len = vsnprintf(temp_buf, 32 * 1024, str, *ap);
-    //vsnprintf returns -1 if the buffer is too small
-    if (temp_len == size_t(-1))
+    std::size_t temp_len = vsnprintf(&temp_buf[0], 32 * 1024, str, *ap);
+    if (temp_len == size_t(-1)) //vsnprintf returns -1 if the buffer is too small
         temp_len = 32*1024-1;
 
-    size_t wtemp_len = 32*1024-1;
-    Utf8toWStr(temp_buf, temp_len, wtemp_buf, wtemp_len);
+    temp_buf.resize(strnlen_s(temp_buf.c_str())); // Resize to match the formatted string
```
Comment (pklloveyou) [inline src/common/Utilities/Util.cpp]: Error showing too few parameters

File: src/common/Utilities/Util.cpp [hunk 3]
```diff
@@ -541,19 +549,24 @@ void utf8printf(FILE* out, const char *str, ...)
...
+
+    if (!temp_buf.empty())
+    {
+        Utf8toWStr(temp_buf, wtemp_buf, 32 * 1024);
```
Comment (pklloveyou) [inline src/common/Utilities/Util.cpp]: Error showing too many parameters

File: src/server/worldserver/CommandLine/CliRunnable.cpp [hunk 4]
```diff
@@ -80,14 +80,21 @@ int cli_hook_func()
 void utf8print(void* /*arg*/, const char* str)
 {
 #if AC_PLATFORM == AC_PLATFORM_WINDOWS
-    wchar_t wtemp_buf[6000];
-    size_t wtemp_len = 6000-1;
-    if (!Utf8toWStr(str, strlen(str), wtemp_buf, wtemp_len))
+    std::wstring wtemp_buf;
+    std::string temp_buf(str);
+
+    if (!Utf8toWStr(temp_buf, wtemp_buf, 6000))
```
Comment (pklloveyou) [inline src/server/worldserver/CommandLine/CliRunnable.cpp]: Error showing too many parameters

File: src/common/Utilities/Util.cpp (hunk 2 above)
Comment (pklloveyou) [reply src/common/Utilities/Util.cpp]: image

**Comment by Viste** [comment]
> Error showing too many parameters

thanks i look

**Comment by pklloveyou** [comment]
```c++
//temp_buf.resize(strnlen_s(temp_buf.c_str()));
temp_buf.resize(strlen(temp_buf.c_str()));
```
What if I do that?

**Comment by Viste** [comment]
> ```c++
> //temp_buf.resize(strnlen_s(temp_buf.c_str()));
> temp_buf.resize(strlen(temp_buf.c_str()));
> ```
> 
> What if I do that?

you dont handle string with 0-terminated ))

**Comment by pklloveyou** [comment]
Oh yes.. I'm a rookie. Thank you for your answer

File: src/server/worldserver/CommandLine/CliRunnable.cpp (hunk 4 above)
Comment (pklloveyou) [inline src/server/worldserver/CommandLine/CliRunnable.cpp]: image

File: src/common/Utilities/Util.cpp (hunk 3 above)
Comment (pklloveyou) [inline src/common/Utilities/Util.cpp]: image

**Comment by Viste** [comment]
@pklloveyou man, keep calm i fix this all in few days just give me some time i have one more task before fix this PR
when I did it, I was in a hurry and I was sleepy, so maybe there are some mistakes. To be clear, I need to sit down and look at the whole code.

**Comment by BarbzYHOOL** [comment]
this looked good

**Pull request closed without merging**